src/ucs/sys/sock.c

```diff
@@ -951,6 +959,87 @@ int ucs_sockaddr_ip_cmp(const struct sockaddr *sa1, const struct sockaddr *sa2)
                   UCS_IPV4_ADDR_LEN : UCS_IPV6_ADDR_LEN);
 }
 
+static void ucs_sockaddr_print_subnet_info(const struct sockaddr *sa1,
+                                           const struct sockaddr *sa2,
+                                           unsigned prefix_len, int matched)
+{
+    UCS_STRING_BUFFER_ONSTACK(info, UCS_SOCKADDR_STRING_LEN * 2 + 60);
+    const struct sockaddr *saddrs[] = {sa1, sa2};
+    unsigned sa_index               = 0;
+    char ip_str[UCS_SOCKADDR_STRING_LEN];
+    const struct sockaddr **sa;
+    ucs_status_t status;
+
+    if (!ucs_log_is_enabled(UCS_LOG_LEVEL_DEBUG)) {
+        return;
+    }
+
+    ucs_string_buffer_appendf(&info, "IP addresses");
+    if (!matched) {
+        ucs_string_buffer_appendf(&info, " do not");
+    }
+
+    ucs_string_buffer_appendf(&info, " match with a %u-bit prefix,",
+                              prefix_len);
+
+    ucs_carray_for_each(sa, saddrs, ucs_static_array_size(saddrs)) {
+        status = ucs_sockaddr_get_ipstr(*sa, ip_str, UCS_SOCKADDR_STRING_LEN);
+        ucs_string_buffer_appendf(&info, " sa%u: %s", ++sa_index,
+                                  (status == UCS_OK) ? ip_str : "<null>");
+    }
+
+    ucs_debug(ucs_string_buffer_cstr(&info));
+}
+
+ucs_status_t ucs_sockaddr_subnet_match(const struct sockaddr *sa1,
+                                       const struct sockaddr *sa2,
+                                       unsigned prefix_len, int *is_match_p)
+{
+    int matched         = 0;
+    ucs_status_t status = UCS_OK;
+    const void *ipaddr1, *ipaddr2;
+    size_t addr_size, addr_size_bits;
+
+    if (!ucs_sockaddr_are_both_known_af(sa1, sa2)) {
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    if (sa1->sa_family != sa2->sa_family) {
+        ucs_debug("different addr_family: s1 %s s2 %s",
+                  ucs_sockaddr_address_family_str(sa1->sa_family),
+                  ucs_sockaddr_address_family_str(sa2->sa_family));
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    /* Get address size */
+    ucs_assert_always(ucs_sockaddr_inet_addr_sizeof(sa1, &addr_size) == UCS_OK);
+    addr_size_bits = addr_size * UCS_SOCKET_BYTE_TO_BITS;
+
+    /* Check subnet mask size (bits belonging to the prefix) */
+    if (prefix_len > addr_size_bits) {
+        ucs_error("prefix is bigger than address size (in bits): prefix=%u"
+                  " size=%lu",
+                  prefix_len, addr_size_bits);
+        status = UCS_ERR_EXCEEDS_LIMIT;
+        goto out;
+    }
+
+    ipaddr1 = ucs_sockaddr_get_inet_addr(sa1);
+    ipaddr2 = ucs_sockaddr_get_inet_addr(sa2);
+    ucs_assertv((ipaddr1 != NULL) && (ipaddr2 != NULL), "ipaddr1=%p ipaddr2=%p",
+                ipaddr1, ipaddr2);
+
+    /* Check if the addresses have matching prefixes */
+    matched = ucs_bitwise_is_equal(ipaddr1, ipaddr2, prefix_len);
+    ucs_sockaddr_print_subnet_info(sa1, sa2, prefix_len, matched);
```

Comment:
What is the rationale in calling a print func when we only compare? I noticed the `if (!ucs_log_is_enabled...` there, and yet...

Reply:
not sure I understand, how does it make difference if we compare or do something else?

Reply:
why print smth when only a comparison is requested? that's werid..

Reply:
for debug purposes